### PR TITLE
Added functions to generate, sign and verify a Json Web Token

### PR DESCRIPTION
**Pull request by ggc87**

### CODE REVIEW

File: emotly/utils.py [hunk 1]
```diff
@@ -43,3 +43,47 @@ def send_email_confirmation(email, confirmation_token):
                                "https://emotly.herokuapp.com/confirm_email/" +
                                confirmation_token)
     message.send()
+
+
+# Generate a Json Web Token fro a given user
```
Comment (michelangelo) [inline emotly/utils.py]: nit: spelling; full stop at the end.

File: emotly/utils.py [hunk 2]
```diff
@@ -43,3 +43,47 @@ def send_email_confirmation(email, confirmation_token):
...
+               'expire': (user.last_login+delta).strftime("%Y:%m:%d %H:%M:%S")
+               }
+    # sign token using a cryptografically hash algoritm
+    signature = sign_jwt(header, payload)
+    jwt = {'header': header, 'payload': payload, 'signature': signature}
+
+    return jwt
+
+
+# Return True if a given token is valid
```
Comment (michelangelo) [inline emotly/utils.py]: nit: full stop at the end.

File: emotly/utils.py [hunk 3]
```diff
@@ -43,3 +43,47 @@ def send_email_confirmation(email, confirmation_token):
...
+def verify_jwt_token(user, token):
+    json_format = json.loads(token)
+    try:
+        User.objects.get(nickname=json_format["payload"]["username"])
```
Comment (michelangelo) [inline emotly/utils.py]: The user might have never confimed his account, hence still be disabled.;)

File: emotly/utils.py [hunk 4]
```diff
@@ -43,3 +43,47 @@ def send_email_confirmation(email, confirmation_token):
...
+    return signature_received == json_format["signature"]
+
+
+# Return a cryptografic hash value used to sign
+# the given message ( header + payload)
+def sign_jwt(header, payload):
+    str_message = str(header)+str(payload)
+    # user a cryptographic hash algorithm to sing header and payload
+    # this "ensure" that the token has been generated from our backend
+    signature = hmac.new(app.config["HMAC_SECRET_KEY"].encode("utf-8"),
```
Comment (michelangelo) [inline emotly/utils.py]: Please, add default value and/or env var for `app.config['HMAC_SECRET_KEY']` in config.py. If you need to add a new env var also add it in the README among the others.

File: emotly/utils.py [hunk 5]
```diff
@@ -43,3 +43,47 @@ def send_email_confirmation(email, confirmation_token):
...
+                         str_message.encode("utf-8"),
+                         hashlib.sha256
+                         ).digest()
```
Comment (michelangelo) [inline emotly/utils.py]: Perhaps you might want to close everything on :86...;)

**Comment by michelangelo** [comment]
Nice stuff! Please address the review comments and ask @tizianasellitto for the second round of review.

Also, please, **do not forget to add coverage** for these functions: they're critically important and we don't want to create regression in future.

Thanks!

File: emotly/utils.py (hunk 3 above)
Comment (michelangelo) [inline emotly/utils.py]: **DISREGARD** this comment, I was wrong: this function only has to deal with the token verification.
User checks should be performed by the callers.

File: emotly/utils.py [hunk 6]
```diff
@@ -43,3 +43,47 @@ def send_email_confirmation(email, confirmation_token):
...
+
+
+# Generate a Json Web Token fro a given user
+# TODO: add support for different hash algorithms
+def gen_jwt_token(user, delta=datetime.timedelta(hours=300)):
+    if user is None:
+        return None
+    # generate the header and the payload message
+    header = {'algo': 'sha256', 'type': 'jwt'}
+    payload = {'username': user.nickname,
```
Comment (tizianasellitto) [inline emotly/utils.py]: Is there a specific reason why it is called "username" instead of "nickname" as in the model?

Comment (ggc87) [inline emotly/utils.py]: no reason. Only my hand writing automatically username ;).I'll change it

File: emotly/utils.py (hunk 3 above)
Comment (ggc87) [inline emotly/utils.py]: I've removed any check on the user.

File: README.md [hunk 7]
```diff
@@ -65,6 +65,7 @@ The following is a (hopefully) updated list of all the env var currently support
 * `EMOTLY_APP_DEBUG_ENABLE` If specified (no matter the value), the app will be run with the `Debug` flag enabled
 * `POSTMARK_API_TOKEN` Token used by Postmark API to send email. Set this to 'POSTMARK_API_TEST' to avoid actually sending mails.
 * `POSTMARK_SENDER` An email used in the from field.
+* `EMOTLY_HMAC_SECRET` [**required**] A string, as complex as possible, for the hmac secret (used for Json Web Token generation)
```
Comment (tizianasellitto) [inline README.md]: It should be "EMOTLY_APP_SEC_HMAC_SECRET" just like the other security related env variable

**Comment by tizianasellitto** [comment]
LGTM.

**Pull request merged**